elfo-core/src/mailbox.rs

```diff
@@ -1,56 +1,166 @@
+    tx_semaphore: Semaphore,
+
+    /// A notifier of a receiver about the availability of new messages.
+    // TODO: replace with `diatomic-waker` (3-5% faster).
+    rx_notify: CachePadded<Notify>,
+
+    /// A trace ID that should be assigned once the mailbox is closed.
+    /// Use `Mutex` here for synchronization on close, more in `close()`.
     closed_trace_id: Mutex<Option<TraceId>>,
 }
 
 impl Mailbox {
     pub(crate) fn new() -> Self {
         Self {
-            queue: GenericChannel::with_capacity(LIMIT),
+            queue: MpscQueue::new_with_stub(Envelope::stub()),
+            tx_semaphore: Semaphore::new(LIMIT),
+            rx_notify: CachePadded::new(Notify::new()),
             closed_trace_id: Mutex::new(None),
         }
     }
 
     pub(crate) async fn send(&self, envelope: Envelope) -> Result<(), SendError<Envelope>> {
-        let fut = self.queue.send(envelope);
-        fut.await.map_err(|err| SendError(err.0))
+        let permit = match self.tx_semaphore.acquire().await {
+            Ok(permit) => permit,
+            Err(_) => return Err(SendError(envelope)),
+        };
+
+        permit.forget();
+        self.queue.enqueue(envelope);
+        self.rx_notify.notify_one();
+        Ok(())
     }
 
     pub(crate) fn try_send(&self, envelope: Envelope) -> Result<(), TrySendError<Envelope>> {
-        self.queue.try_send(envelope).map_err(|err| match err {
-            channel::TrySendError::Full(envelope) => TrySendError::Full(envelope),
-            channel::TrySendError::Closed(envelope) => TrySendError::Closed(envelope),
-        })
+        match self.tx_semaphore.try_acquire() {
+            Ok(permit) => {
+                permit.forget();
+                self.queue.enqueue(envelope);
+                self.rx_notify.notify_one();
+                Ok(())
+            }
+            Err(TryAcquireError::NoPermits) => Err(TrySendError::Full(envelope)),
+            Err(TryAcquireError::Closed) => Err(TrySendError::Closed(envelope)),
+        }
     }
 
     pub(crate) async fn recv(&self) -> RecvResult {
-        let fut = self.queue.receive();
-        match fut.await {
-            Some(envelope) => RecvResult::Data(envelope),
-            None => self.on_close(),
+        loop {
+            // TODO: it should be possible to use `dequeue_unchecked()` here.
+            // Preliminarily, we should guarantee that `recv()` can be called only
+            // by one consumer. However, it's not enough to create a dedicated
+            // `MailboxConsumer` because users can steal `Context` to another
+            // task/thread and create a race with the `drop_all()` method.
+            if let Some(envelope) = self.queue.dequeue() {
+                self.tx_semaphore.add_permits(1);
+                return RecvResult::Data(envelope);
+            }
+
+            if self.tx_semaphore.is_closed() {
+                return self.on_close();
+            }
+
+            self.rx_notify.notified().await;
```

Comment:
I’m not sure this is correct in presence of multiple concurrent consumers (which are currently possible per the comment above).

Consider this timeline (queue is empty at the start):
**Receiver 1:** tries and fails to `.dequeue()`
**Receiver 2:** tries and fails to `.dequeue()`
**Sender:** queues a message
**Sender:** queues another message
**Receiver 1:** reaches `.notified()`, which consumes the notification and returns
**Receiver 1:** takes one message from the queue, leaving one queued
**Receiver 2:** reaches `.notified()` and hangs, despite there being a message in the queue

I’m not sure whether this case is something we’re concerned about, but it seems possible.

Reply:
Multiple concurrent consumers are possible only in case of incorrect context usage when an actor is already done, but the context is still used from another thread. I'm not sure we should guarantee the absence of hangs here, only soundness.

Reply:
It seems like concurrent `.recv()` is in fact impossible even in theory, so this race can’t ever happen.